common/src/character_controller/vector_bounds.rs

```diff
@@ -0,0 +1,298 @@
+        // difficulty is finding which set of linearly independent bounds need to be applied so that all bounds are
+        // satisfied. Since bounds are one-sided and not guaranteed to be linearly independent from each other, this
+        // requires some ad-hoc choices. The algorithm we choose here is to (1) assume that `new_bound` is one of these
+        // linearly independent bounds, (2) if necessary, pair it up with each existing bound to find the first such
+        // bound that allows all bounds to be satisfied, and (3) zero out the vector if no such pairing works, as we
+        // assume that we need to apply three linearly independent bounds.
+
+        // Apply new_bound if necessary.
+        if !new_bound.check_vector(vector, self.error_margin) {
+            new_bound.constrain_vector(vector, self.error_margin);
+            if let Some(ref mut tagalong) = tagalong {
+                // Note: The tagalong vector does not need an error margin.
+                new_bound.constrain_vector(tagalong, 0.0);
+            }
+        }
+
+        // Check if all constraints are satisfied
+        if (self.bounds.iter()).all(|b| b.check_vector(vector, self.error_margin)) {
+            return;
+        }
+
+        // If not all constraints are satisfied, find the first constraint that if applied will satisfy
+        // the remaining constriants
+        for bound in (self.bounds.iter()).filter(|b| !b.check_vector(vector, self.error_margin)) {
+            let Some(ortho_bound) = bound.get_self_constrained_with_bound(new_bound)
+            else {
+                warn!("Unsatisfied existing bound is parallel to new bound. Is the character squeezed between two walls?");
+                continue;
+            };
+
+            let mut candidate = *vector;
+            ortho_bound.constrain_vector(&mut candidate, self.error_margin);
+
+            if (self.bounds.iter()).all(|b| b.check_vector(&candidate, self.error_margin)) {
+                *vector = candidate;
+                if let Some(ref mut tagalong) = tagalong {
+                    ortho_bound.constrain_vector(tagalong, 0.0);
+                }
+                return;
+            }
+        }
+
+        // If no choice satisfies all constraints, keep all bounds and set the vector to 0
+        vector.set_zero();
+        if let Some(ref mut tagalong) = tagalong {
+            tagalong.set_zero();
+        }
+    }
+}
+
+/// Represents a single constraint for a vector. `VectorBound`s alone conceptually contain
+/// enough information to apply to a vector, but practically, one other piece of information
+/// is needed: `error_margin`, which exists in `VectorBoundGroup`.
+#[derive(Clone)]
+pub struct VectorBound {
+    normal: na::UnitVector3<f32>,
+    projection_direction: na::UnitVector3<f32>,
+    error_margin_factor: f32, // Margin of error when the bound is applied
```

Comment:
Not in this commit. I probably should have waited until the "Add logic to prevent slanted walls from lifting the player up" commit before adding this multiplier feature. After that commit, the possible values are 1.0 and -1.0.

The logic in that commit is probably one of the most difficult-to-follow pieces of logic in the entire PR. I've made quite an effort to make it more follow-able, but it's definitely not perfect.

Reply:
I've replaced this with a boolean, since that makes the logic much easier to follow in my opinion. Thanks for calling this out!